Framework/Beamline/test/ComponentInfoTest.h

```diff
@@ -14,18 +16,192 @@ class ComponentInfoTest : public CxxTest::TestSuite {
+
+    TS_ASSERT_THROWS(
+        ComponentInfo(
+            bankSortedDetectorIndices, ranges, positions, rotations,
+            boost::make_shared<DetectorInfo>() /*Detector info size 0*/),
+        std::invalid_argument &);
+  }
+
+  void test_throw_if_positions_rotation_inputs_different_sizes() {
+    std::vector<size_t> detectorIndices{}; // No detectors in this example
+    std::vector<std::pair<size_t, size_t>> ranges;
+    ranges.push_back(std::make_pair(0, 0)); // One component with no detectors
+    auto positions = boost::make_shared<std::vector<Eigen::Vector3d>>(
+        1); // 1 position provided
+    auto rotations = boost::make_shared<std::vector<Eigen::Quaterniond>>(
+        0); // 0 rotations provided
+
+    TS_ASSERT_THROWS(ComponentInfo(detectorIndices, ranges, positions,
+                                   rotations,
+                                   boost::make_shared<DetectorInfo>()),
+                     std::invalid_argument &);
+  }
+
+  void test_throw_if_positions_and_rotations_not_same_size_as_ranges() {
+    /*
+     * Positions are rotations are only currently stored for non-detector
+     * components
+     * We should have as many ranges as we have non-detector components too.
+     * All vectors should be the same size.
+     */
+    std::vector<size_t> detectorIndices{}; // No detectors in this example
+    std::vector<std::pair<size_t, size_t>> ranges; // Empty ranges!
+    auto positions = boost::make_shared<std::vector<Eigen::Vector3d>>(
+        1); // 1 position provided
+    auto rotations = boost::make_shared<std::vector<Eigen::Quaterniond>>(
+        1); // 1 rotation provided
+
+    TS_ASSERT_THROWS(ComponentInfo(detectorIndices, ranges, positions,
+                                   rotations,
+                                   boost::make_shared<DetectorInfo>()),
+                     std::invalid_argument &);
+  }
+
+  void test_read_positions_rotations() {
+
+    /*
+           |
+     ------------
+     |         | 1
+    -------
+    | 0  | 2
+    */
+
+    std::vector<Eigen::Vector3d> detPositions;
+    detPositions.emplace_back(1, 0, 0);
+    detPositions.emplace_back(2, 0, 0);
+    detPositions.emplace_back(3, 0, 0);
+    std::vector<Eigen::Quaterniond> detRotations;
+    detRotations.emplace_back(
+        Eigen::AngleAxisd(M_PI / 5, Eigen::Vector3d::UnitZ()));
+    detRotations.emplace_back(
+        Eigen::AngleAxisd(M_PI / 4, Eigen::Vector3d::UnitZ()));
+    detRotations.emplace_back(
+        Eigen::AngleAxisd(M_PI / 3, Eigen::Vector3d::UnitZ()));
+    auto detectorInfo =
+        boost::make_shared<DetectorInfo>(detPositions, detRotations);
+    std::vector<size_t> bankSortedDetectorIndices{0, 2, 1};
+
+    std::vector<std::pair<size_t, size_t>> ranges;
+    ranges.push_back(std::make_pair(0, 3));
+    ranges.push_back(std::make_pair(0, 2));
+
+    auto compPositions = boost::make_shared<std::vector<Eigen::Vector3d>>();
```

Comment:
By having `shared_ptr` for the constructor arguments in `ComponentInfo` you get a lot of cumbersome client code. Consider passing a vector (by value or `&&`) and change the visitor getters to move its members when accessed for construction of `ComponentInfo` (i.e., you can get the positions from the visitor only once). Maybe having a one-shot factory method on this visitor to create the `ComponentInfo`?